File: examples/python/multi_vector_example.py
#!/usr/bin/env python
"""
Multi-Vector Embeddings Example for EngramDB

This example demonstrates how to use multi-vector embeddings (ColBERT/ColPali-style)
in EngramDB to perform more precise semantic search through late interaction matching.
"""

import engramdb
import time

def main():
    print("=== EngramDB Multi-Vector Embeddings Example (Python) ===")
    
    # Create an embedding service with multi-vector capability
    # For this example, we'll use the mock providers
    service = engramdb.EmbeddingService.new_mock_multi_vector(96, 20)
    
    print(f"Created embedding service with multi-vector capability")
    print(f"Single-vector dimensions: {service.dimensions()}")
    print(f"Multi-vector dimensions: {service.multi_vector_dimensions()}")
    print(f"Has multi-vector capability: {service.has_multi_vector()}")
    
    # Create example documents
    documents = [
        "EngramDB is a specialized database for agent memory systems.",
        "It supports both single-vector and multi-vector embeddings.",
        "Multi-vector embeddings allow for more precise similarity matching.",
        "The late interaction approach compares individual token embeddings.",
        "ColBERT and ColPali are examples of multi-vector embedding models.",
    ]
    
    # Create both single-vector and multi-vector nodes
    single_vector_nodes = []
    multi_vector_nodes = []
    
    print("\nCreating memory nodes with both embedding types:")
    for i, doc in enumerate(documents):
        # Create a single-vector node
        node_sv = engramdb.MemoryNode.from_text(doc, service, None)
        single_vector_nodes.append(node_sv)
        
        # Create a multi-vector node
        node_mv = engramdb.MemoryNode.from_text_multi_vector(doc, service, None)
        multi_vector_nodes.append(node_mv)
        
        print(f"- Document {i+1}: Created single and multi-vector nodes")
        
        sv_dims = len(node_sv.embeddings_legacy()) if node_sv.embeddings() is not None else 0
        mv_dims = (f"{node_mv.multi_vector_embeddings().num_vectors()} vectors of "
                  f"{node_mv.multi_vector_embeddings().dimensions()} dimensions each")
        
        print(f"  └ Single vector dimensions: {sv_dims}")
        print(f"  └ Multi vector: {mv_dims}")
    
    # Create a database with multi-vector support
    print("\nCreating database with multi-vector index")
    
    # Create a database configured for multi-vector
    db_config = engramdb.DatabaseConfig.new_in_memory()
    db_config.set_vector_algorithm(engramdb.VectorAlgorithm.MultiVector)
    
    # Set multi-vector config with late interaction similarity
    mv_config = engramdb.MultiVectorIndexConfig()
    mv_config.set_similarity_method(engramdb.MultiVectorSimilarityMethod.LateInteraction)
    db_config.set_multi_vector_config(mv_config)
    
    db = engramdb.Database.with_config(db_config)
    
    # Add multi-vector nodes to the database
    print("Adding multi-vector nodes to the database...")
    for node in multi_vector_nodes:
        db.add(node)
    
    # Perform a multi-vector search
    query = "Tell me about vector embeddings"
    print(f"\nSearching for: \"{query}\"")
    
    # Generate multi-vector query embeddings
    query_multi_vec = service.generate_multi_vector_for_query(query)
    print(f"Query has {query_multi_vec.num_vectors()} vectors of {query_multi_vec.dimensions()} dimensions")
    
    # Search with the specialized multi-vector index
    print("\nPerforming search...")
    start_time = time.time()
    
    # We need to use the underlying vector from multi-vector for the API
    # This is a compatibility layer - internally it will use the multi-vector similarity
    query_vec = query_multi_vec.vectors()[0]
    results = db.search_by_vector(query_vec, 3, 0.0)
    
    search_time = time.time() - start_time
    print(f"Search completed in {search_time:.4f} seconds")
    
    print("\nResults:")
    for i, (node, score) in enumerate(results):
        content = node.get_attribute("content").as_string()
        print(f"{i+1}. Score: {score:.4f} - \"{content}\"")
    
    print("\nExample completed successfully!")

if __name__ == "__main__":
    main()

File: examples/rust/multi_vector_example.rs
use engramdb::{
    core::{MemoryNode, AttributeValue},
    database::{Database, DatabaseConfig},
    StorageType,
    embeddings::{
        EmbeddingService, MockMultiVectorProvider,
        EmbeddingError,
    },
    vector::{
        MultiVectorIndex, MultiVectorIndexConfig, MultiVectorSimilarityMethod,
        VectorAlgorithm, VectorIndexConfig,
    },
    error::EngramDbError
};
use std::fmt;

// Custom error type that can handle both EngramDbError and EmbeddingError
#[derive(Debug)]
enum ExampleError {
    DatabaseError(EngramDbError),
    EmbeddingError(EmbeddingError),
}

impl fmt::Display for ExampleError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            ExampleError::DatabaseError(e) => write!(f, "Database error: {}", e),
            ExampleError::EmbeddingError(e) => write!(f, "Embedding error: {}", e),
        }
    }
}

impl std::error::Error for ExampleError {}

impl From<EngramDbError> for ExampleError {
    fn from(err: EngramDbError) -> Self {
        ExampleError::DatabaseError(err)
    }
}

impl From<EmbeddingError> for ExampleError {
    fn from(err: EmbeddingError) -> Self {
        ExampleError::EmbeddingError(err)
    }
}

impl From<Box<dyn std::error::Error>> for ExampleError {
    fn from(err: Box<dyn std::error::Error>) -> Self {
        ExampleError::EmbeddingError(EmbeddingError::GenerationFailed(err.to_string()))
    }
}

fn main() -> std::result::Result<(), ExampleError> {
    println!("=== EngramDB Multi-Vector Embeddings Example ===");

    // Create a mock embedding service with multi-vector capability
    let mut service = EmbeddingService::new_mock(96);
    let multi_provider = MockMultiVectorProvider::new(96, 20);
    service.add_multi_vector_provider(multi_provider);

    println!("Created embedding service with multi-vector capability");
    println!("Dimensions: {}", service.dimensions());
    println!("Multi-vector dimensions: {:?}", service.multi_vector_dimensions());

    // Create example documents
    let documents = vec![
        "EngramDB is a specialized database for agent memory systems.",
        "It supports both single-vector and multi-vector embeddings.",
        "Multi-vector embeddings allow for more precise similarity matching.",
        "The late interaction approach compares individual token embeddings.",
        "ColBERT and ColPali are examples of multi-vector embedding models.",
    ];

    // Create both single-vector and multi-vector nodes
    let mut single_vector_nodes = Vec::new();
    let mut multi_vector_nodes = Vec::new();

    println!("\nCreating memory nodes with both embedding types:");
    for (i, doc) in documents.iter().enumerate() {
        // Create a single-vector node
        let node_sv = MemoryNode::from_text(doc, &service, None)?;
        
        // Get dimensions before moving
        let sv_dimensions = node_sv.embeddings().map(|v| v.len()).unwrap_or(0);
        single_vector_nodes.push(node_sv);

        // Create a multi-vector node
        let multi_vec = service.generate_multi_vector_for_document(doc, None)?;
        let node_mv = MemoryNode::with_multi_vector(multi_vec);
        
        // Get dimensions before moving
        let mv_num_vectors = node_mv.multi_vector_embeddings().map(|mv| mv.num_vectors()).unwrap_or(0);
        let mv_dimensions = node_mv.multi_vector_embeddings().map(|mv| mv.dimensions()).unwrap_or(0);
        
        multi_vector_nodes.push(node_mv);

        println!("- Document {}: Created single and multi-vector nodes", i + 1);
        println!("  └ Single vector dimensions: {}", sv_dimensions);
        println!("  └ Multi vector: {} vectors of {} dimensions each", 
            mv_num_vectors, mv_dimensions);
    }

    // Create a specialized multi-vector index
    let mut mv_index = MultiVectorIndex::with_config(MultiVectorIndexConfig {
        quantize: false,
        similarity_method: MultiVectorSimilarityMethod::LateInteraction,
    });

    println!("\nAdding multi-vector nodes to specialized index");
    for node in &multi_vector_nodes {
        mv_index.add(node)?;
    }

    // Perform a multi-vector search
    let query = "Tell me about vector embeddings";
    let query_multi_vec = service.generate_multi_vector_for_query(query)?;

    println!("\nSearching for: \"{}\"", query);
    println!("Query has {} vectors", query_multi_vec.num_vectors());

    let results = mv_index.search(&query_multi_vec, 3, 0.0)?;
    
    println!("\nResults using late interaction scoring:");
    for (i, (id, score)) in results.iter().enumerate() {
        let node = multi_vector_nodes.iter().find(|n| n.id() == *id).unwrap();
        let content = node.get_attribute("content")
.and_then(|attr| match attr {
                AttributeValue::String(s) => Some(s.as_str()),
                _ => None,
            })
            .unwrap_or_default();
        
        println!("{}. Score: {:.4} - \"{}\"", i + 1, score, content);
    }

    // Compare with other similarity methods
    println!("\nComparing different similarity methods:");
    
    // MaxSim
    let config_max = MultiVectorIndexConfig {
        quantize: false,
        similarity_method: MultiVectorSimilarityMethod::Maximum,
    };
    let mut index_max = MultiVectorIndex::with_config(config_max);
    for node in &multi_vector_nodes {
        index_max.add(node)?;
    }
    let results_max = index_max.search(&query_multi_vec, 3, 0.0)?;
    
    println!("\nResults using maximum similarity:");
    for (i, (id, score)) in results_max.iter().enumerate() {
        let node = multi_vector_nodes.iter().find(|n| n.id() == *id).unwrap();
        let content = node.get_attribute("content")
.and_then(|attr| match attr {
                AttributeValue::String(s) => Some(s.as_str()),
                _ => None,
            })
            .unwrap_or_default();
        
        println!("{}. Score: {:.4} - \"{}\"", i + 1, score, content);
    }
    
    // Average
    let config_avg = MultiVectorIndexConfig {
        quantize: false,
        similarity_method: MultiVectorSimilarityMethod::Average,
    };
    let mut index_avg = MultiVectorIndex::with_config(config_avg);
    for node in &multi_vector_nodes {
        index_avg.add(node)?;
    }
    let results_avg = index_avg.search(&query_multi_vec, 3, 0.0)?;
    
    println!("\nResults using average similarity:");
    for (i, (id, score)) in results_avg.iter().enumerate() {
        let node = multi_vector_nodes.iter().find(|n| n.id() == *id).unwrap();
        let content = node.get_attribute("content")
.and_then(|attr| match attr {
                AttributeValue::String(s) => Some(s.as_str()),
                _ => None,
            })
            .unwrap_or_default();
        
        println!("{}. Score: {:.4} - \"{}\"", i + 1, score, content);
    }

    // Create a database with multi-vector support
    println!("\nCreating database with multi-vector index");
    let vector_config = VectorIndexConfig {
        algorithm: VectorAlgorithm::MultiVector,
        hnsw: None,
        multi_vector: Some(MultiVectorIndexConfig::default()),
    };
    
    let db_config = DatabaseConfig {
        storage_type: StorageType::Memory,
        storage_path: None,
        cache_size: 100,
        vector_index_config: vector_config,
    };
    
    let _db = Database::new(db_config)?; // Database is unused in this example

    // Since database doesn't expose direct access to vector_index, we'll use our own
    // MultiVectorIndex here and skip using the Database
    
    println!("Using separate MultiVectorIndex to perform search");
    println!("Added {} multi-vector nodes to index", multi_vector_nodes.len());
    
    // Use direct MultiVectorIndex instead of going through Database
    let mut direct_index = MultiVectorIndex::new();
    
    // Add the nodes to the index
    for node in &multi_vector_nodes {
        direct_index.add(node)?;
    }
    
    // Perform a query using the direct index
    let results = direct_index.search(&query_multi_vec, 3, 0.0)?;
    
    println!("\nDatabase search results:");
    for (i, (id, score)) in results.iter().enumerate() {
        // Get the node from our local list since the Database interface doesn't expose get_by_id
        let node = multi_vector_nodes.iter().find(|n| n.id() == *id).unwrap();
        let content = node.get_attribute("content")
.and_then(|attr| match attr {
                AttributeValue::String(s) => Some(s.as_str()),
                _ => None,
            })
            .unwrap_or_default();
        
        println!("{}. Score: {:.4} - \"{}\"", i + 1, score, content);
    }

    println!("\nExample completed successfully!");
    Ok(())
}